File: sdks/python/dcp_ai/v2/crypto_registry.py
"""
Algorithm registry for pluggable cryptographic providers.
"""

from __future__ import annotations

import threading

from dcp_ai.v2.crypto_provider import CryptoProvider, KemProvider


class AlgorithmRegistry:
    """Thread-safe registry mapping algorithm names to provider instances."""

    def __init__(self) -> None:
        self._signers: dict[str, CryptoProvider] = {}
        self._kems: dict[str, KemProvider] = {}

    def register_signer(self, provider: CryptoProvider) -> None:
        self._signers[provider.alg] = provider

    def register_kem(self, provider: KemProvider) -> None:
        self._kems[provider.alg] = provider

    def get_signer(self, alg: str) -> CryptoProvider:
        try:
            return self._signers[alg]
        except KeyError:
            raise KeyError(f"No signer registered for algorithm: {alg}")

    def get_kem(self, alg: str) -> KemProvider:
        try:
            return self._kems[alg]
        except KeyError:
            raise KeyError(f"No KEM registered for algorithm: {alg}")

    def has_signer(self, alg: str) -> bool:
        return alg in self._signers

    def has_kem(self, alg: str) -> bool:
        return alg in self._kems

    def list_signers(self) -> list[str]:
        return list(self._signers.keys())

    def list_kems(self) -> list[str]:
        return list(self._kems.keys())


_default_registry: AlgorithmRegistry | None = None
_registry_lock = threading.Lock()


def get_default_registry() -> AlgorithmRegistry:
    """Return the module-level singleton AlgorithmRegistry, creating it on first access."""
    global _default_registry
    if _default_registry is None:
        with _registry_lock:
            if _default_registry is None:
                _default_registry = AlgorithmRegistry()
    return _default_registry

File: sdks/go/dcp/v2/crypto_registry.go
package v2

import "sort"

// AlgorithmRegistry manages registered signature and KEM providers.
type AlgorithmRegistry struct {
	sigProviders map[string]CryptoProvider
	kemProviders map[string]KemProvider
}

// NewAlgorithmRegistry creates an empty registry.
func NewAlgorithmRegistry() *AlgorithmRegistry {
	return &AlgorithmRegistry{
		sigProviders: make(map[string]CryptoProvider),
		kemProviders: make(map[string]KemProvider),
	}
}

// RegisterSigner adds a signature provider, keyed by its Alg().
func (r *AlgorithmRegistry) RegisterSigner(p CryptoProvider) {
	r.sigProviders[p.Alg()] = p
}

// RegisterKem adds a KEM provider, keyed by its Alg().
func (r *AlgorithmRegistry) RegisterKem(p KemProvider) {
	r.kemProviders[p.Alg()] = p
}

// GetSigner returns the signature provider for the given algorithm.
func (r *AlgorithmRegistry) GetSigner(alg string) (CryptoProvider, error) {
	p, ok := r.sigProviders[alg]
	if !ok {
		return nil, &ErrUnknownAlgorithm{Alg: alg, Kind: "signer"}
	}
	return p, nil
}

// GetKem returns the KEM provider for the given algorithm.
func (r *AlgorithmRegistry) GetKem(alg string) (KemProvider, error) {
	p, ok := r.kemProviders[alg]
	if !ok {
		return nil, &ErrUnknownAlgorithm{Alg: alg, Kind: "kem"}
	}
	return p, nil
}

// HasSigner reports whether a signer is registered for the given algorithm.
func (r *AlgorithmRegistry) HasSigner(alg string) bool {
	_, ok := r.sigProviders[alg]
	return ok
}

// HasKem reports whether a KEM is registered for the given algorithm.
func (r *AlgorithmRegistry) HasKem(alg string) bool {
	_, ok := r.kemProviders[alg]
	return ok
}

// ListSigners returns sorted algorithm names for all registered signers.
func (r *AlgorithmRegistry) ListSigners() []string {
	names := make([]string, 0, len(r.sigProviders))
	for k := range r.sigProviders {
		names = append(names, k)
	}
	sort.Strings(names)
	return names
}

// ListKems returns sorted algorithm names for all registered KEMs.
func (r *AlgorithmRegistry) ListKems() []string {
	names := make([]string, 0, len(r.kemProviders))
	for k := range r.kemProviders {
		names = append(names, k)
	}
	sort.Strings(names)
	return names
}
